remove index param from enableManualSummarization

daesun-park opened the pull request:
## Description

This PR fixes a bug, where the manual summarization currently does not work when the `TestTreeProvider` creates multiple trees.

daesun-park committed:
remove index param from enableManualSummarization

CraigMacomber commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -113,9 +113,9 @@ export class TestTreeProvider {
-        assert(index < this.trees.length, "Index out of bounds: not enough trees");
-        const summarizer = await createSummarizer(this.provider, this.containers[index]);
+    public async enableManualSummarization(): Promise<() => Promise<void>> {
+        assert(this.trees.length === 1, "Manual summarization cannot be enabled for multiple trees.")
```
It would be good to document what the issue preventing this is (ex: a todo comment like `// TODO: support manual summarization with multiple trees once bug #???? is fixed.`.)

daesun-park replied:
`enableManualSummarization()` method removed

CraigMacomber commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -113,9 +113,9 @@ export class TestTreeProvider {
      * @param index - the container that will spawn the summarizer client
```
this documentation seems out of date.

daesun-park committed:
PR review comments

daesun-park committed:
Merge branch 'main' into testTreeProvider-summarization-bug

daesun-park committed:
prettier fix

CraigMacomber commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -113,12 +113,15 @@ export class TestTreeProvider {
-        assert(index < this.trees.length, "Index out of bounds: not enough trees");
-        const summarizer = await createSummarizer(this.provider, this.containers[index]);
+    public async enableManualSummarization(): Promise<() => Promise<void>> {
+        // TODO: support manual summarization with multiple trees once bug #1876 is fixed.
```
Issue #1876 is already closed.

If instead you mean https://dev.azure.com/fluidframework/internal/_workitems/edit/1876 in our internal tracker thats a problem since it is not accessible to many people who might look at this.

Also, I'm not sure thats the right item to reference here or that item's status is wrong. Its status indicates that it's in review (meaning that someone, presumable you, has a fix for that bug in PR). If that really was the case, we should probably wait for that bug to be fixed (and thus manual summarization with multiple trees would work) and not do this PR at all.

If we are going to work around a bug without fixing it, we should not track the workaround as if it's fixing the bug: it should get its own work item.

If we are going to leave a known bug in the codebase, it should be documented where all users of the codebase can see, meaning on github (as an issue and/or documentation in the codebase) and not just our internal tracker.

CraigMacomber replied:
If we are going to require "This may only be called when exactly one tree has been created and has not submitted any edits", I'd like to code to document why this is the case. If its a bug we can explain that. If its not a bug, we should still have some explanation. Avoiding the whole issue as @noencke proposes by moving this to an option of create sounds like a great idea.

CraigMacomber replied:
And thanks for the details @noencke : that helps a lot!

yann-achard-MS replied:
> and you assert that if `summarizeOnDemand` is true then `trees === 1`

Wouldn't it be possible to support `trees !== 0` instead? You'd create the first tree, then make it the summarizer, then create the rest of the trees (if any).

yann-achard-MS commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -113,12 +113,15 @@ export class TestTreeProvider {
+        // TODO: support manual summarization with multiple trees once bug #1876 is fixed.
+        assert(
+            this.trees.length === 1,
+            "Manual summarization cannot be enabled for multiple trees.",
```
It's still fine to have multiple trees if you create those additional trees *after* calling `enableManualSummarization`, right?
Assuming that's correct (and it seems to be since I'm relying on it), we should mention that in the message. Otherwise you get the wrong idea as to what is really possible or not.

daesun-park replied:
Changes made such that the `create()` method now takes in a `summarizeOnDemand` parameter, and creates the summarizer after creating the first tree (subsequent trees are created afterwards).

daesun-park committed:
enable manual summary during tree creation.

daesun-park committed:
prettier formating

noencke commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -83,11 +84,14 @@ export class TestTreeProvider {
-    public static async create(trees = 0): Promise<ITestTreeProvider> {
+    public static async create(trees = 0, summarizeOnDemand = false): Promise<ITestTreeProvider> {
         const provider = new TestTreeProvider() as ITestTreeProvider;
         for (let i = 0; i < trees; i++) {
```
I think the optimal logic here is:
1. Assert trees !== 0 || summarizeOnDemand is false
2. Create the first tree if trees > 0
3. Create the summarizer if summarizeOnDemand is true. Save the summarizer in a field
4. Create the rest of the trees

This implies replacing the summarizeOnDemand field with the summarizer itself. Then the manualSummarize() method should assert that this.summarizer !== undefined before summarizing.

daesun-park replied:
Sounds good Noah :)

However, if we create the summarizer within the for loop (immediately after the first tree), will we still need the assert statement?
```
    public static async create(trees = 0, summarizeOnDemand = false): Promise<ITestTreeProvider> {
        const provider = new TestTreeProvider() as ITestTreeProvider;
        for (let i = 0; i < trees; i++) {
            await provider.createTree();
            if (summarizeOnDemand && i === 0) {
                const summarizer = await createSummarizer(provider, provider.containers[0]);
                this.summarizer = async () => {
                    await summarizeNow(summarizer, "TestTreeProvider");
                };
            }
        }

        return provider;
    }
```

noencke replied:
Yes. The way you wrote it above is elegant (nice work!) and it does indeed skip the problematic case. However, I'm of the opinion that we should prefer throwing errors over silently "failing". For example, right now, if a user calls `summarize()`, and the summarizer is undefined, we do nothing. It would be better for a user if we instead right away threw an error that said "can't summarize because... " (this is _especially_ true because this is a test utility. Sometimes in production code it's preferable to fail silently rather than throwing errors so that a user doesn't end up seeing an error message, but in test helper code we can be aggressive about emitting errors because we want to be sure that the developer is using the utilities correctly!). So, what do we put after the "because..."? Well, `summarize` could either fail because 'summarizeOnDemand' was never set to true (oops! Let's tell the dev to set it to true!). Or, it could fail because, although they set `summarizeOnDemand` to true, they set `trees` to 0. In that case, we want to give them a different error message ("must create at least one tree"). We can't tell inside the `summarize` method which of these two cases we're in (the summarizer is undefined but we don't know why), but if we instead move the second error to happen right away in `create()`, we dodge that problem. And anyway, it is usually best to error as soon as possible so that the dev can fix their code a quickly as possible. We know that if they call `create(0, true)` then they must not understand the contract. Let's educate them right away with an error rather than letting it slide and letting them remain confused :)

daesun-park replied:
Two assert statements in `create` and `summarize` methods added.
create:
```
        assert(
            !(trees === 0 && summarizeOnDemand),
            "can't summarize, because must create at least one tree.",
        );
```
summarizer:
```
        assert(
            this.summarizer !== undefined,
            "can't summarize because summarizeOnDemand was not set to true.",
        );
```

noencke commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -109,6 +113,19 @@ export class TestTreeProvider {
+     * set to true.
+     * @returns void after a summary has been resolved. May be called multiple times.
+     */
+    public async manualSummarize(): Promise<void> {
```
How about just "summarize"?

noencke commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -109,6 +113,19 @@ export class TestTreeProvider {
+     */
+    public async manualSummarize(): Promise<void> {
+        if (TestTreeProvider.summarizeOnDemand) {
+            const summarizer = await createSummarizer(this.provider, this.containers[0]);
```
Note that this creates another summarizer every time this is called; we only want to make one summarizer if we make one at all, which is why it would do well as an optional field as described above.

daesun-park replied:
Summarizer is now created only once in the `create` method

daesun-park committed:
fixed logic for creating summarizer

daesun-park commented:
I have updated the `create()` method to set the summarizer (as a class field), and updated the test cases in `testTreeProvider.spec.ts` to use the new `summarize()` method.

As we no longer require the `enableManualSummarization` method, should I go ahead and remove it? I did notice that it's also being used in `SharedTree.spec.ts`, so should I update the test cases to use the new `summarize()` method there as well?

noencke commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -109,16 +119,31 @@ export class TestTreeProvider {
+     * @returns void after a summary has been resolved. May be called multiple times.
+     */
+    public async summarize(): Promise<void> {
+        if (TestTreeProvider.summarizer !== undefined) {
```
This should throw an error if the summarizer is undefined, rather than silently doing nothing (see related comment in `create()`

daesun-park replied:
Added assert statement to throw error when summarizer is undefined

noencke commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -61,6 +63,7 @@ export class TestTreeProvider {
     private readonly provider: ITestObjectProvider;
     private readonly _trees: ISharedTree[] = [];
     private readonly _containers: IContainer[] = [];
+    public static summarizer: Summarizer;
```
`create()` is `static`, but this field can be non-static. Statics can be dangerous (e.g. what happens if we run tests in parallel and they share the same summarizer across tests?). 

To do this, you'll need to write a constructor for this class that takes a summarizer. I don't think constructors can be async, so it will also need to take the first tree. So you'll want two overloads, something like:

```
constructor(tree?: ISharedTree);
constructor(tree: ISharedTree, summarizer: Summarizer);
constructor(tree?: ISharedTree, summarizer?: Summarizer) {
    ...
}

Unfortunately, this also means a bunch of the code that is currently inside of the constructor will need to be lifted up into the `create()` method, since we need to be able to create the first tree before constructing, and the constructor will in fact require another parameter, namely the `TestObjectProvider` (the `provider field`).

I'm sorry this is getting more complicated than I thought it would. I think it's still very doable, but if you would like some help just let me know and we can pair program a bit.

daesun-park replied:
Updated `constructor` and `create` methods to create the first tree within `create`, and summarizer field initialized in the `constructor`

noencke commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -83,11 +86,18 @@ export class TestTreeProvider {
             await provider.createTree();
+            if (summarizeOnDemand && i === 0) {
+                const summarizer = await createSummarizer(provider, provider.containers[0]);
+                this.summarizer = async () => {
```
I think this code would be a little more clear if `this.summarizer` was the result of `createSummarizer` (the 'summarizer' on line 94) rather than an async function. Then, the body of 'summarize()' can call 'summarizeNow' instead, and we can remove this little function altogether.

daesun-park replied:
Updated `this.summarizer` to be the result of `createSummarizer()`, and called in the `summarize` method

daesun-park committed:
updated constructor to take in summarizer

daesun-park committed:
changed constructor parameters

daesun-park committed:
remove enableManualSummarization method

daesun-park committed:
fixed comment

noencke commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -111,37 +151,34 @@ export class TestTreeProvider {
+        if (firstTreeParams !== undefined) {
+            const [container, firstTree, summarizer] = firstTreeParams;
+            this._containers.push(container);
+            this._trees[this.trees.length] = firstTree;
```
use `push` here. It's written this funny way on line 147 because it's being used as a return value there.

daesun-park replied:
Updated to use push

noencke reviewed (commented):
Looks great - this last round of feedback is mostly just cosmetic. Great work! The code is clean and you added some good tests.

daesun-park committed:
PR review changes

daesun-park committed:
removed unused Summarizer type

daesun-park committed:
fixed test cases

daesun-park commented on packages/dds/tree/src/test/util/testTreeProvider.spec.ts:
```diff
@@ -8,18 +8,53 @@ import { SharedTreeCore } from "../../shared-tree-core";
+        } catch (error) {
+            summarizerError = error;
+        }
+        assert(summarizerError !== undefined);
```
I tried to use `assert.throws`, but it seems to ask me to provide the expected error object to be thrown. I tried to manually create it by calling the `AssertionError()` function, but this did not seem to work. The parameters I passed in were the fields I found for the error object in the `try` and `catch` statement from the current test.

Do you happen to know if there is a better way to get the exact `AssertionError` object?

My attempt:
```
        const summarizerError = new AssertionError({
            actual: false,
            expected: true,
            operator:'==',
            message: "can't summarize, because must create at least one tree."
        });
        const provider = await TestTreeProvider.create(1);
        assert.throws(
            async () => provider.summarize(),
            summarizerError
        )
```

noencke replied:
You don't have to provide a second argument if you don't want. If you want to check that it's the right kind of error though, search our packages for `assert.throws` and you'll see some examples, e.g. the one in utilities.spec.ts

daesun-park commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -61,6 +64,7 @@ export class TestTreeProvider {
     private readonly provider: ITestObjectProvider;
     private readonly _trees: ISharedTree[] = [];
     private readonly _containers: IContainer[] = [];
+    public readonly summarizer: ISummarizer | undefined;
```
Updated to `summarizer?: ISummarizer`, and made field private.

daesun-park commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -83,12 +87,48 @@ export class TestTreeProvider {
+    public static async create(trees = 0, summarizeOnDemand = false): Promise<ITestTreeProvider> {
+        assert(
+            !(trees === 0 && summarizeOnDemand),
+            "can't summarize, because must create at least one tree.",
```
Updated assert message

daesun-park commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -83,12 +87,48 @@ export class TestTreeProvider {
-        for (let i = 0; i < trees; i++) {
-            await provider.createTree();
+    public static async create(trees = 0, summarizeOnDemand = false): Promise<ITestTreeProvider> {
+        assert(
```
added comment

daesun-park committed:
changed equal to strictEqual

daesun-park committed:
removed unused comment
